File: packages/quran-engine-py/quran_engine/names_depth.py
"""The 99 Names of Allah in depth. Mirrors src/namesDepth.js.

`names-of-allah.json` carries each name, its meaning and a one-line description. This carries
what sits beneath that: the triliteral root the name is built on, one of nine themes, a
paragraph on what the name means, one line on living by it, and the ayahs where the name
itself appears with the token it sits at.

Roots print SPACED ("ر ح م") the way a grammar book sets them, while morphology.json stores them
closed up ("رحم"). `root_key` closes the spaces, so matching a name to its morphology entries is
one call rather than a trap.

The written material is Tilawa's (Jamil Hammoudeh), used with permission; the occurrences point
into this engine's own Hafs text.
"""
from __future__ import annotations
from typing import Optional


class NameDepth(dict):
    """number, root, theme, explanation, living, occurrences."""


class NameTheme(dict):
    """id, label."""


def root_key(root: str) -> str:
    """A spaced root as morphology stores it: "ر ح م" -> "رحم"."""
    return "".join((root or "").split())


class NamesDepth:
    def __init__(self, data: Optional[dict] = None):
        data = data or {}
        self._names = [NameDepth(n) for n in data.get("names", [])]
        self._themes = [NameTheme(t) for t in data.get("themes", [])]
        self._by_number = {n["number"]: n for n in self._names}

    @property
    def is_loaded(self) -> bool:
        return bool(self._names)

    def all(self) -> list[NameDepth]:
        """All 99, ordered by number."""
        return self._names

    def by_number(self, number: int) -> Optional[NameDepth]:
        return self._by_number.get(number)

    def themes(self) -> list[NameTheme]:
        """The nine themes, in the corpus's own order."""
        return self._themes

    def theme(self, theme_id: str) -> Optional[NameTheme]:
        return next((t for t in self._themes if t["id"] == theme_id), None)

    def by_theme(self, theme_id: str) -> list[NameDepth]:
        return [n for n in self._names if n.get("theme") == theme_id]

    def by_root(self, root: str) -> list[NameDepth]:
        """Every name built on one root. Accepts either spelling, spaced or closed up."""
        key = root_key(root)
        if not key:
            return []
        return [n for n in self._names if root_key(n.get("root", "")) == key]

    def in_ayah(self, surah_id: int, ayah_id: int) -> list[dict]:
        """Every name appearing in an ayah, with the occurrence that put it there.

        An occurrence whose `token` is None could not be placed in the ayah's tokens (ten of
        them, across four names); the ayah is still right, so it sorts last rather than first.
        """
        hits = []
        for name in self._names:
            for occurrence in name.get("occurrences", []):
                if occurrence["surah"] == surah_id and occurrence["ayah"] == ayah_id:
                    hits.append({"name": name, "occurrence": occurrence})
        return sorted(hits, key=lambda h: (h["occurrence"]["token"] is None, h["occurrence"]["token"] or 0))

    def search(self, query: str, limit: int = 25) -> list[NameDepth]:
        """Names whose root, explanation or living line matches a query."""
        q = (query or "").strip()
        if not q:
            return []
        lower = q.lower()
        key = root_key(q)
        out = [
            n for n in self._names
            if (key and key in root_key(n.get("root", "")))
            or lower in n.get("explanation", "").lower()
            or lower in n.get("living", "").lower()
        ]
        return out[:limit]

    def count(self) -> dict:
        return {
            "names": len(self._names),
            "themes": len(self._themes),
            "occurrences": sum(len(n.get("occurrences", [])) for n in self._names),
        }

File: packages/quran-engine-js/src/namesDepth.js
// @ts-check
/**
 * The 99 Names of Allah in depth: the layer under `namesOfAllah`.
 *
 * `names-of-allah.json` carries each name, its meaning and a one-line description. This carries
 * what sits beneath that: the triliteral root the name is built on, one of nine themes, a
 * paragraph on what the name means, one line on living by it, and the ayahs where the name
 * itself appears with the token it sits at.
 *
 * Roots print SPACED ("ر ح م") the way a grammar book sets them, while `morphology.json` stores
 * them closed up ("رحم"). `rootKey()` closes the spaces, so matching a name to its morphology
 * entries is one call rather than a trap.
 *
 * The written material (roots, themes, explanations, living lines) is Tilawa's, by Jamil
 * Hammoudeh, used with permission. The occurrences point into this engine's own Ḥafṣ text.
 *
 * @typedef {Object} NameTheme
 * @property {string} id
 * @property {string} label
 *
 * @typedef {Object} NameOccurrence
 * @property {number} surah
 * @property {number} ayah
 * @property {number|null} token  0-based whitespace-token index into the ayah's raw text, or
 *   null where the corpus could not place the name in the ayah (ten occurrences across four
 *   names). The ayah is still right: show the verse whole and highlight nothing.
 * @property {number} tokens  how many tokens the name spans (0 for an unplaced occurrence)
 *
 * @typedef {Object} NameDepth
 * @property {number} number       1..99, matching `namesOfAllah.byNumber`
 * @property {string} root         spaced, e.g. "ر ح م"
 * @property {string} theme        a `NameTheme` id
 * @property {string} explanation
 * @property {string} living
 * @property {NameOccurrence[]} occurrences
 */

/** A spaced root as morphology stores it: "ر ح م" -> "رحم". */
export function rootKey(root) {
  return String(root ?? "").replace(/\s+/gu, "");
}

export class NamesDepth {
  /**
   * @param {Object} [data] parsed data/names-depth.json
   * @param {NameTheme[]} [data.themes]
   * @param {NameDepth[]} [data.names]
   */
  constructor(data = {}) {
    /** @type {NameDepth[]} */
    this._names = data.names ?? [];
    /** @type {NameTheme[]} */
    this._themes = data.themes ?? [];
    /** @type {Map<number, NameDepth>} */
    this._byNumber = new Map(this._names.map((n) => [n.number, n]));
  }

  get isLoaded() {
    return this._names.length > 0;
  }

  /** All 99, ordered by number. @returns {NameDepth[]} */
  all() {
    return this._names;
  }

  /** @param {number} number 1..99 @returns {NameDepth|null} */
  byNumber(number) {
    return this._byNumber.get(number) ?? null;
  }

  /** The nine themes, in the corpus's own order. @returns {NameTheme[]} */
  themes() {
    return this._themes;
  }

  /** @param {string} id a theme id @returns {NameTheme|null} */
  theme(id) {
    return this._themes.find((t) => t.id === id) ?? null;
  }

  /** Every name under one theme, by number. @param {string} id @returns {NameDepth[]} */
  byTheme(id) {
    return this._names.filter((n) => n.theme === id);
  }

  /**
   * Every name built on one root. Accepts either spelling, spaced or closed up.
   * @param {string} root @returns {NameDepth[]}
   */
  byRoot(root) {
    const key = rootKey(root);
    if (!key) return [];
    return this._names.filter((n) => rootKey(n.root) === key);
  }

  /**
   * Every name that appears in an ayah, with the occurrence that put it there.
   * @param {number} surahId @param {number} ayahId
   * @returns {{name: NameDepth, occurrence: NameOccurrence}[]}
   */
  inAyah(surahId, ayahId) {
    /** @type {{name: NameDepth, occurrence: NameOccurrence}[]} */
    const hits = [];
    for (const name of this._names) {
      for (const occurrence of name.occurrences) {
        if (occurrence.surah === surahId && occurrence.ayah === ayahId) hits.push({ name, occurrence });
      }
    }
    // Unplaced occurrences (null token) sort last, so a highlighted list stays in reading order.
    const at = (o) => (o.token ?? Number.MAX_SAFE_INTEGER);
    return hits.sort((a, b) => at(a.occurrence) - at(b.occurrence));
  }

  /**
   * Names whose root, explanation or living line matches a query. The Arabic root is compared as
   * written (spaces closed on both sides); the prose case-insensitively.
   * @param {string} query @param {number} [limit] @returns {NameDepth[]}
   */
  search(query, limit = 25) {
    const q = String(query ?? "").trim();
    if (!q) return [];
    const lower = q.toLowerCase();
    const key = rootKey(q);
    return this._names
      .filter((n) => (key && rootKey(n.root).includes(key))
        || n.explanation.toLowerCase().includes(lower)
        || n.living.toLowerCase().includes(lower))
      .slice(0, limit);
  }

  count() {
    return {
      names: this._names.length,
      themes: this._themes.length,
      occurrences: this._names.reduce((n, x) => n + x.occurrences.length, 0),
    };
  }
}
